db-service/lib/deep-queries.js

```diff
@@ -195,7 +203,10 @@ const getDeepQueries = (query, dbData, target) => {
     diff = [diff]
   }
 
-  return _getDeepQueries(diff, target, true)
+  const deletes = new Map()
```

Comment:
I like the idea, but it shows a clear problem in the overall implementation. As the `INSERT` map is not re-used and rather generated after each `_getDeepQueries` call. So from my perspective it would be best if `DELETE` and `INSERT` both use the same approach and when taking this approach it should probably also just include `UPDATE` and return all of them as separate results as this would also remove the additional split on the called of `getDeepQueries`.

Reply:
Yes, we could do that for `INSERT`, too

Reply:
I just noticed that it's a little bit more subtle:
There's a query hierarchy which needs to be taken into account w.r.t. managed fields. If a given entity has sub queries (i.e. queries for its children), it is considered 'modified' and the managed fields (`@cds.on.update`) must be updated. That is currently wrong in this PR (deleting children wouldn't count as 'modified').

I think the simplest solution would be to keep the query hierarchy intact and collect same-target queries afterwards, as it's already done for `INSERT`.

A more performant solution would be to introduce a 'dirty' flag to mark modified children. I'll think about this.

Reply:
prototype solution is https://github.com/cap-js/cds-dbs/blob/fff3ed46f42016c9a5c7d440289ebcaf7391e206/db-service/lib/deep-queries.js#L215

(still need to fix a bug with root insert result)